.ci/scripts/test_old_deps.sh

```diff
@@ -9,12 +12,74 @@ set -ex
 
 apt-get update
 apt-get install -y \
-        python3 python3-dev python3-pip python3-venv \
-        libxml2-dev libxslt-dev xmlsec1 zlib1g-dev tox libjpeg-dev libwebp-dev
+        python3 python3-dev python3-pip python3-venv pipx \
+        libxml2-dev libxslt-dev xmlsec1 zlib1g-dev libjpeg-dev libwebp-dev
 
 export LANG="C.UTF-8"
 
 # Prevent virtualenv from auto-updating pip to an incompatible version
 export VIRTUALENV_NO_DOWNLOAD=1
 
-exec tox -e py3-old
+# Patch the project definitions in-place:
+# - Replace all lower and tilde bounds with exact bounds
+# - Make the pyopenssl 17.0, which is the oldest version that works with
+#   a `cryptography` compiled against OpenSSL 1.1.
+# - Delete all lines referring to psycopg2 --- so no testing of postgres support.
+# - Omit systemd: we're not logging to journal here.
+
+# TODO: we should also replace caret bounds, see
+#    https://python-poetry.org/docs/dependency-specification/#version-constraints
+# We don't use these yet, but they are the default bound used when you `poetry add` from
+# the commandline, rather than editing pyproject.toml directly.
+# The sed expression 's/\^/==/g' ought to do the trick. But it would also change
+# `python = "^3.7"` to `python = "==3.7"`, which would mean we fail because olddeps
+# runs on 3.8 (#12343).
+
+# TODO: I'd prefer to use something like this
+#   https://github.com/python-poetry/poetry/issues/3527
+#   https://github.com/pypa/pip/issues/8085
+# rather than this sed script.
+
+sed -i-backup \
```

Comment:
is there any point in keeping a backup here?

```suggestion
sed -i \
```

Reply:
Argh, I could have sworn I already removed this somewhere else.

(A previous version of this script diffed the pre- and post patch toml. Unfortunately `toml.dump`'s output style is different to our input, so the resulting diff isn't very meaningful.)